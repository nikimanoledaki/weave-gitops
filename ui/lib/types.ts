import { Condition, FluxObjectKind, Interval } from "./api/core/types.pb";

export enum PageRoute {
  Applications = "/applications",
  ApplicationDetail = "/application_detail",
  ApplicationAdd = "/application_add",
  ApplicationRemove = "/application_remove",
  GitlabOAuthCallback = "/oauth/gitlab",
}

export enum GrpcErrorCodes {
  Unknown = 2,
  Unauthenticated = 16,
  NotFound = 5,
}

export type RequestError = Error & {
  code?: number;
};

export enum V2Routes {
  Automations = "/applications",
  Sources = "/sources",
  FluxRuntime = "/flux_runtime",
  Kustomization = "/kustomization",
  HelmRelease = "/helm_release",
  HelmRepo = "/helm_repo",
  GitRepo = "/git_repo",
  HelmChart = "/helm_chart",
  Bucket = "/bucket",

  // Use this to allow for certain components to route to a 404 and still compile.
  // We want to keep certain components around for future use.
  NotImplemented = "/not_implemented",
}

export const WeGONamespace = "flux-system";
export const DefaultCluster = "Default";
export const NoNamespace = "";

export interface Source {
  name?: string;
  namespace?: string;
  kind?: FluxObjectKind;
  conditions?: Condition[];
  interval?: Interval;
  suspended?: boolean;
  clusterName?: string;
  lastUpdatedAt?: string;
}

export interface Syncable {
  name?: string;
  kind?: FluxObjectKind;
  namespace?: string;
  clusterName?: string;
}
