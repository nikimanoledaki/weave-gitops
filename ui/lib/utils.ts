import _ from "lodash";
import { toast } from "react-toastify";
import { computeReady } from "../components/KubeStatusIndicator";
import { Condition, HelmRelease, Kustomization } from "./api/core/types.pb";
import { PageRoute } from "./types";

export function notifySuccess(message: string) {
  toast["success"](message);
}

export function notifyError(message: string) {
  toast["error"](`Error: ${message}`);
}

// Must be one of the valid URLs that we have already
// configured on the Gitlab backend for our Oauth app.
export function gitlabOAuthRedirectURI(): string {
  return `${window.location.origin}${PageRoute.GitlabOAuthCallback}`;
}

export function poller(cb, interval): any {
  if (process.env.NODE_ENV === "test") {
    // Stay synchronous in tests
    return cb();
  }

  return setInterval(cb, interval);
}

export function isHTTP(uri: string): boolean {
  // Regex from Diego Perini's gist: https://gist.github.com/dperini/729294
  // It works better than other regular expressions for validating HTTP and HTTPS URLs.
  const regex = new RegExp(
    /^(?:(?:(?:https?|ftp):)?\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z0-9\u00a1-\uffff][a-z0-9\u00a1-\uffff_-]{0,62})?[a-z0-9\u00a1-\uffff]\.)+(?:[a-z\u00a1-\uffff]{2,}\.?))(?::\d{2,5})?(?:[/?#]\S*)?$/i
  );

  return regex.test(uri);
}

export function convertGitURLToGitProvider(uri: string): string {
  if (isHTTP(uri)) {
    return uri;
  }

  const matches = uri.match(/git@(.*)[/|:](.*)\/(.*)/);
  if (!matches) {
    throw new Error(`could not parse url "${uri}"`);
  }
  const [, provider, org, repo] = matches;

  return `https://${provider}/${org}/${repo}`;
}

export function pageTitleWithAppName(title: string, appName?: string): string {
  return `${title}${appName ? ` for ${appName}` : ""}`;
}

interface Statusable {
  conditions: Condition[];
  suspended: boolean;
}

export function statusSortHelper({
  suspended,
  conditions,
}: Statusable): number {
  if (suspended) return 2;
  if (computeReady(conditions)) return 3;
  else return 1;
}

export function automationLastUpdated(a: Kustomization | HelmRelease): string {
  return _.get(_.find(a?.conditions, { type: "Ready" }), "timestamp");
}

const kindPrefix = "Kind";

export function addKind(kind: string): string {
  if (!kind.startsWith(kindPrefix)) {
    return `${kindPrefix}${kind}`;
  }
  return kind;
}

export function removeKind(kind: string): string {
  if (kind.startsWith(kindPrefix)) {
    return kind.slice(kindPrefix.length);
  }
  return kind;
}

export function makeImageString(images: string[]): string {
  let imageString = "";
  if (!images[0]) return "-";
  else imageString += images[0];
  if (images[1]) {
    for (let i = 1; i < images.length; i++) imageString += `\n${images[i]}`;
  }
  return imageString;
}

export function formatMetadataKey(key: string): string {
  return key
    .replace(/-/g, " ")
    .replace(/\w+/g, (w) => w[0].toUpperCase() + w.slice(1));
}
