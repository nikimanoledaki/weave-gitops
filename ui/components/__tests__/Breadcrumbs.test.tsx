import "jest-styled-components";
import React from "react";
import renderer from "react-test-renderer";
import { createMockClient, withContext, withTheme } from "../../lib/test-utils";
import Breadcrumbs from "../Breadcrumbs";

describe("Breadcrumbs", () => {
  describe("snapshots", () => {
    it("renders", () => {
      const tree = renderer
        .create(
          withTheme(
            withContext(<Breadcrumbs />, "/applications", {
              applicationsClient: createMockClient({}),
            })
          )
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("renders child route", () => {
      const tree = renderer
        .create(
          withTheme(
            withContext(<Breadcrumbs />, "/kustomization?name=flux", {
              applicationsClient: createMockClient({}),
            })
          )
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("renders on the root page", () => {
      const tree = renderer
        .create(
          withTheme(
            withContext(<Breadcrumbs />, "/", {
              applicationsClient: createMockClient({}),
            })
          )
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
