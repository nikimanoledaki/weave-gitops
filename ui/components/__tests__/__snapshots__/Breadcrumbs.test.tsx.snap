// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots renders 1`] = `
<div
  className="sc-bczRLJ iDpicH"
>
  <a
    className="sc-eCYdqJ sc-jSMfEi kYfFWP bYfIap"
    href="/applications"
    onClick={[Function]}
  >
    <span
      className="sc-gsnTZi hkopcK"
      color="primary"
      size="medium"
    >
      Applications
    </span>
  </a>
</div>
`;

exports[`snapshots renders child route 1`] = `
<div
  className="sc-bczRLJ iDpicH"
>
  <a
    className="sc-eCYdqJ sc-jSMfEi kYfFWP bYfIap"
    href="/applications"
    onClick={[Function]}
  >
    <span
      className="sc-gsnTZi hkopcK"
      color="primary"
      size="medium"
    >
      Applications
    </span>
  </a>
  <div
    className="sc-bczRLJ iDpicH sc-dkzDqf fphFew"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
    </svg>
  </div>
  <a
    className="sc-eCYdqJ sc-jSMfEi kYfFWP bYfIap"
    href="/kustomization?name=flux"
    onClick={[Function]}
  >
    <span
      className="sc-gsnTZi cZFkGq"
      color="primary"
      size="medium"
    >
      flux
    </span>
  </a>
</div>
`;

exports[`snapshots renders on the root page 1`] = `
<div
  className="sc-bczRLJ iDpicH"
>
  <a
    className="sc-eCYdqJ sc-jSMfEi kYfFWP bYfIap"
    href="/"
    onClick={[Function]}
  >
    <span
      className="sc-gsnTZi hkopcK"
      color="primary"
      size="medium"
    />
  </a>
</div>
`;
